File: javascript/packages/browser/test/visitor.test.ts
import { describe, test, expect, beforeAll } from "vitest"
import { Herb, Visitor } from "../src"
import type { Node, HTMLTextNode } from "../src/index.js"

class RecordingVisitor extends Visitor {
  visited: string[] = []

  visitChildNodes(node: Node): void {
    this.visited.push(node.constructor.name)
    super.visitChildNodes(node)
  }
}

class TextNodeVisitor extends Visitor {
  textNodes: string[] = []

  visitHTMLTextNode(node: HTMLTextNode) {
    this.textNodes.push(node.content)
  }
}

describe("Visitor", () => {
  beforeAll(async () => {
    await Herb.load()
  })

  test("traverses nodes", () => {
    const visitor = new RecordingVisitor()

    const result = Herb.parse("<p>Hello</p>");
    result.visit(visitor)

    expect(visitor.visited).toEqual([
      "DocumentNode",
      "HTMLElementNode",
      "HTMLOpenTagNode",
      "HTMLTextNode",
      "HTMLCloseTagNode",
    ])
  })

  test("text content visitor", () => {
    const visitor = new TextNodeVisitor()

    const result = Herb.parse("<p>Hello</p>");
    result.visit(visitor)

    expect(visitor.textNodes).toEqual([
      "Hello"
    ])
  })
})

File: test/visitor_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class VisitorTest < Minitest::Spec
  class VisitedNodesVisitor < Herb::Visitor
    attr_reader :visited_nodes

    def initialize
      super
      @visited_nodes = []
    end

    def visit_child_nodes(node)
      @visited_nodes << node
      super
    end
  end

  test "visitor" do
    visitor = VisitedNodesVisitor.new

    result = Herb.parse(%(<p id="greeting">Hello <%= user.name %></p>))
    result.visit(visitor)

    expected_nodes = [
      "Herb::AST::DocumentNode",
      "Herb::AST::HTMLElementNode",
      "Herb::AST::HTMLOpenTagNode",
      "Herb::AST::HTMLAttributeNode",
      "Herb::AST::HTMLAttributeNameNode",
      "Herb::AST::LiteralNode",
      "Herb::AST::HTMLAttributeValueNode",
      "Herb::AST::LiteralNode",
      "Herb::AST::HTMLTextNode",
      "Herb::AST::ERBContentNode",
      "Herb::AST::HTMLCloseTagNode"
    ]

    assert result.success?
    assert_equal expected_nodes, visitor.visited_nodes.map(&:class).map(&:to_s)
  end

  test "document with nil in child_nodes" do
    visitor = VisitedNodesVisitor.new

    result = Herb.parse(%(<p>Hello))
    result.visit(visitor)

    expected_nodes = [
      "Herb::AST::DocumentNode",
      "Herb::AST::HTMLOpenTagNode",
      "Herb::AST::HTMLTextNode"
    ]

    assert result.failed?
    assert_equal expected_nodes, visitor.visited_nodes.map(&:class).map(&:to_s)
  end
end
